### PR TITLE
Test central functions in phonebook backend.

### PR DESCRIPTION
**Pull request by agger-magenta**
Test retrieval of information from MO, compare with output known to be correct.

### CODE REVIEW

File: backend/import/requirements.txt [hunk 1]
```diff
@@ -1,7 +1,19 @@
+atomicwrites==1.2.1
+attrs==18.2.0
 cached-property==1.5.1
 certifi==2018.10.15
 chardet==3.0.4
+flake8==3.6.0
```
Comment (danchr) [inline backend/import/requirements.txt]: Usually, test requirements get put into a separate file called `requirements-text.txt` or `requirements-dev.txt`…

File: backend/import/do_import.py [hunk 2]
```diff
@@ -161,6 +159,7 @@ def writer(data):
     try:
         write_phonebook_data(orgunit_writer, employee_writer)
     except Exception as e:
+        raise
```
Comment (danchr) [inline backend/import/do_import.py]: Leftover debugging?

File: backend/import/test_mo_api.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    employee_uuid = '0014209a-8832-44b3-b7a0-4cc209a71993'
+    ou_uuid = '0418617c-242f-4d9a-81cc-abb269ad27b4'
+
+    ou = mo_api.MOOrgUnit(ou_uuid)
+    ou.get = my_mo_get
+    e = mo_api.MOEmployee(employee_uuid)
+    e.get = my_mo_get
+
+    # Test json function
+    assert(ou.json['name'] == 'Østervrå børnehus')
```
Comment (danchr) [inline backend/import/test_mo_api.py]: `assert` is a keyword statement, like `print` used to be, not a function. The parentheses are redundant in this case.

File: backend/import/.test_data/mo/http:##morademo.atlas.magenta.dk#service#e#0014209a-8832-44b3-b7a0-4cc209a71993#details#.json [hunk 4]
```diff
@@ -0,0 +1 @@
+{"address": true, "association": false, "engagement": true, "it": false, "leave": false, "manager": false, "org_unit": false, "role": false}
```
Comment (danchr) [inline backend/import/.test_data/mo/http:##morademo.atlas.magenta.dk#service#e#0014209a-8832-44b3-b7a0-4cc209a71993#details#.json]: You should probably for these files, to make them ever so slightly less unreadable. `python -m json.tool` is nice.

File: backend/run_tests.sh [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+# Run phonebook backend tests.
+
+cd $(dirname $0)/import
+
+venv/bin/pytest --flake8
```
Comment (danchr) [inline backend/run_tests.sh]: You'll want to add a Jenkinsfile for running the tests, and make it generate XML output to feed back to Jenkins. Also, coverage data is nice!

Comment (agger-magenta) [reply backend/run_tests.sh]: Good idea!

File: backend/import/.test_data/mo/http:##morademo.atlas.magenta.dk#service#e#0014209a-8832-44b3-b7a0-4cc209a71993#details#.json (hunk 4 above)
Comment (agger-magenta) [reply backend/import/.test_data/mo/http:##morademo.atlas.magenta.dk#service#e#0014209a-8832-44b3-b7a0-4cc209a71993#details#.json]: The intent is, though, not that they should be read by humans but that they should store actual results of `request.get().json()`. `json.dump(...)` may indeed be too barebones, though.

File: backend/import/test_mo_api.py (hunk 3 above)
Comment (agger-magenta) [reply backend/import/test_mo_api.py]: OK, will fix. Redundancy is evil. :-)

File: backend/import/do_import.py (hunk 2 above)
Comment (agger-magenta) [reply backend/import/do_import.py]: Removed.

File: backend/run_tests.sh (hunk 5 above)
Comment (cmoesgaard) [reply backend/run_tests.sh]: I made a pretty bare-bones one for `flask_saml_sso` that might be useful for this project as well.
https://github.com/magenta-aps/flask_saml_sso/blob/development/Jenkinsfile

File: backend/import/.test_data/mo/http:##morademo.atlas.magenta.dk#service#e#0014209a-8832-44b3-b7a0-4cc209a71993#details#.json (hunk 4 above)
Comment (danchr) [reply backend/import/.test_data/mo/http:##morademo.atlas.magenta.dk#service#e#0014209a-8832-44b3-b7a0-4cc209a71993#details#.json]:
```python
json.dumps(obj, indent=2)
```

This is, essentially, what `python -m json.tool` does…

File: backend/import/test_mo_api.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+import json
+
+import mo_api
+import do_import
+
+
+def my_mo_get(url):
+    url_file = '.test_data/mo/' + url.replace('/', '#') + '.json'
```
Comment (danchr) [inline backend/import/test_mo_api.py]: I can't help but find this is semi-evil — using `requests_mock` seems simpler, and I'd probably elect to combine the various fixtures into one. That's a matter of taste, though, not a complain _as such_…

File: backend/import/test_mo_api.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def test_get_orgunit_data():
+    """Test the get_orgunit_data function in the do_import module."""
+    orgunit_uuid = '0418617c-242f-4d9a-81cc-abb269ad27b4'
+    ou = mo_api.MOOrgUnit(orgunit_uuid)
+    ou.get = my_mo_get
+    ou_data = do_import.get_orgunit_data(ou)
+    # Normalize
+    ou_data = json.loads(json.dumps(ou_data))
```
Comment (danchr) [inline backend/import/test_mo_api.py]:
```suggestion
    ou_data = json.loads(json.dumps(ou_data), indent=2)
```

I like readability!

File: backend/import/test_mo_api.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        assert e_data == target_e_data
+
+
+def test_get_orgunit_data():
+    """Test the get_orgunit_data function in the do_import module."""
+    orgunit_uuid = '0418617c-242f-4d9a-81cc-abb269ad27b4'
+    ou = mo_api.MOOrgUnit(orgunit_uuid)
+    ou.get = my_mo_get
+    ou_data = do_import.get_orgunit_data(ou)
+    # Normalize
```
Comment (danchr) [inline backend/import/test_mo_api.py]: You should probably mention tuples here — I was stumped for a while!

File: backend/import/test_mo_api.py (hunk 6 above)
Comment (agger-magenta) [reply backend/import/test_mo_api.py]: We'll get to that in a later iteration, I think this works for now.

**Pull request merged**